### PR TITLE
nydus-image: add `backend-config` support to `nydus-image unpack`

### PR DESCRIPTION
**Pull request by loheagn**
The `nydus-image unpack` only support reading nydus blob from the blob files on local machine (by the argument `blob`) by now. This patch adds a new argument `backend-config` to this command to allow `nydus-image unpack` to read blob data from all kinds of backends.

### CODE REVIEW

**Comment by jiangliu** [comment]
How about using "-C, --config <config>" instead? We are going to switch from "--backend-config" to "--config" and "--backend-config" will get deprecated.

**Comment by loheagn** [comment]
> How about using "-C, --config " instead? We are going to switch from "--backend-config" to "--config" and "--backend-config" will get deprecated.

@jiangliu But I think this will be conflict with the `backend` in the `Config` when load rafs. https://github.com/dragonflyoss/image-service/blob/master/rafs/src/metadata/mod.rs#L685

**Comment by jiangliu** [comment]
> > How about using "-C, --config " instead? We are going to switch from "--backend-config" to "--config" and "--backend-config" will get deprecated.
> 
> @jiangliu But I think this will be conflict with the `backend` in the `Config` when load rafs. https://github.com/dragonflyoss/image-service/blob/master/rafs/src/metadata/mod.rs#L685

Are you going to support this usage scenario?
1) nydus meta blob has already been extracted to a local directory
2) nydus data blobs haven't been downloaded to the local directory

**Comment by loheagn** [comment]
> > > How about using "-C, --config " instead? We are going to switch from "--backend-config" to "--config" and "--backend-config" will get deprecated.
> > 
> > 
> > @jiangliu But I think this will be conflict with the `backend` in the `Config` when load rafs. https://github.com/dragonflyoss/image-service/blob/master/rafs/src/metadata/mod.rs#L685
> 
> Are you going to support this usage scenario?
> 
> 1. nydus meta blob has already been extracted to a local directory
> 2. nydus data blobs haven't been downloaded to the local directory

Yes. 

For example, by now, when we use the `coverter.Unpack` method in nydus-snapshotter to unpack a nydus image layer, we need to extract both the bootstrap part and the blob data part to the local disk first to allow the nydus-image to read to blob data (use the localfs backend). But with this patch, we can just need to extract the bootstrap part and set up a local http server to serve the blob data part. Then the `backend-config` can be used to pass the information about this local http server.

**Comment by jiangliu** [comment]
> > > > How about using "-C, --config " instead? We are going to switch from "--backend-config" to "--config" and "--backend-config" will get deprecated.
> > > 
> > > 
> > > @jiangliu But I think this will be conflict with the `backend` in the `Config` when load rafs. https://github.com/dragonflyoss/image-service/blob/master/rafs/src/metadata/mod.rs#L685
> > 
> > 
> > Are you going to support this usage scenario?
> > 
> > 1. nydus meta blob has already been extracted to a local directory
> > 2. nydus data blobs haven't been downloaded to the local directory
> 
> Yes.
> 
> For example, by now, when we use the `coverter.Unpack` method in nydus-snapshotter to unpack a nydus image layer, we need to extract both the bootstrap part and the blob data part to the local disk first to allow the nydus-image to read to blob data (use the localfs backend). But with this patch, we can just need to extract the bootstrap part and set up a local http server to serve the blob data part. Then the `backend-config` can be used to pass the information about this local http server.

Can we let nydusd to handle the bootstrap part too?

**Comment by loheagn** [comment]
> > > > > How about using "-C, --config " instead? We are going to switch from "--backend-config" to "--config" and "--backend-config" will get deprecated.
> > > > 
> > > > 
> > > > @jiangliu But I think this will be conflict with the `backend` in the `Config` when load rafs. https://github.com/dragonflyoss/image-service/blob/master/rafs/src/metadata/mod.rs#L685
> > > 
> > > 
> > > Are you going to support this usage scenario?
> > > 
> > > 1. nydus meta blob has already been extracted to a local directory
> > > 2. nydus data blobs haven't been downloaded to the local directory
> > 
> > 
> > Yes.
> > For example, by now, when we use the `coverter.Unpack` method in nydus-snapshotter to unpack a nydus image layer, we need to extract both the bootstrap part and the blob data part to the local disk first to allow the nydus-image to read to blob data (use the localfs backend). But with this patch, we can just need to extract the bootstrap part and set up a local http server to serve the blob data part. Then the `backend-config` can be used to pass the information about this local http server.
> 
> Can we let nydus-image to handle the bootstrap part too? That's a feature I expect, then nydus-image can process remote images.



> > > > > How about using "-C, --config " instead? We are going to switch from "--backend-config" to "--config" and "--backend-config" will get deprecated.
> > > > 
> > > > 
> > > > @jiangliu But I think this will be conflict with the `backend` in the `Config` when load rafs. https://github.com/dragonflyoss/image-service/blob/master/rafs/src/metadata/mod.rs#L685
> > > 
> > > 
> > > Are you going to support this usage scenario?
> > > 
> > > 1. nydus meta blob has already been extracted to a local directory
> > > 2. nydus data blobs haven't been downloaded to the local directory
> > 
> > 
> > Yes.
> > For example, by now, when we use the `coverter.Unpack` method in nydus-snapshotter to unpack a nydus image layer, we need to extract both the bootstrap part and the blob data part to the local disk first to allow the nydus-image to read to blob data (use the localfs backend). But with this patch, we can just need to extract the bootstrap part and set up a local http server to serve the blob data part. Then the `backend-config` can be used to pass the information about this local http server.
> 
> Can we let nydus-image to handle the bootstrap part too? That's a feature I expect, then nydus-image can process remote images.

@jiangliu Yes, I think we can. 

We need to implement the logic of extracting the bootstrap and blob data from the whole nydus image layer (which is implemented in nydus-snapshotter by now) and let the caller tell nydus-image the loacation of the image layer by some argument (may be `--image-path`).

What's your option? @imeoer

**Comment by jiangliu** [comment]
> > > > > > How about using "-C, --config " instead? We are going to switch from "--backend-config" to "--config" and "--backend-config" will get deprecated.
> > > > > 
> > > > > 
> > > > > @jiangliu But I think this will be conflict with the `backend` in the `Config` when load rafs. https://github.com/dragonflyoss/image-service/blob/master/rafs/src/metadata/mod.rs#L685
> > > > 
> > > > 
> > > > Are you going to support this usage scenario?
> > > > 
> > > > 1. nydus meta blob has already been extracted to a local directory
> > > > 2. nydus data blobs haven't been downloaded to the local directory
> > > 
> > > 
> > > Yes.
> > > For example, by now, when we use the `coverter.Unpack` method in nydus-snapshotter to unpack a nydus image layer, we need to extract both the bootstrap part and the blob data part to the local disk first to allow the nydus-image to read to blob data (use the localfs backend). But with this patch, we can just need to extract the bootstrap part and set up a local http server to serve the blob data part. Then the `backend-config` can be used to pass the information about this local http server.
> > 
> > 
> > Can we let nydus-image to handle the bootstrap part too? That's a feature I expect, then nydus-image can process remote images.
> 
> > > > > > How about using "-C, --config " instead? We are going to switch from "--backend-config" to "--config" and "--backend-config" will get deprecated.
> > > > > 
> > > > > 
> > > > > @jiangliu But I think this will be conflict with the `backend` in the `Config` when load rafs. https://github.com/dragonflyoss/image-service/blob/master/rafs/src/metadata/mod.rs#L685
> > > > 
> > > > 
> > > > Are you going to support this usage scenario?
> > > > 
> > > > 1. nydus meta blob has already been extracted to a local directory
> > > > 2. nydus data blobs haven't been downloaded to the local directory
> > > 
> > > 
> > > Yes.
> > > For example, by now, when we use the `coverter.Unpack` method in nydus-snapshotter to unpack a nydus image layer, we need to extract both the bootstrap part and the blob data part to the local disk first to allow the nydus-image to read to blob data (use the localfs backend). But with this patch, we can just need to extract the bootstrap part and set up a local http server to serve the blob data part. Then the `backend-config` can be used to pass the information about this local http server.
> > 
> > 
> > Can we let nydus-image to handle the bootstrap part too? That's a feature I expect, then nydus-image can process remote images.
> 
> @jiangliu Yes, I think we can.
> 
> We need to implement the logic of extracting the bootstrap and blob data from the whole nydus image layer (which is implemented in nydus-snapshotter by now) and let the caller tell nydus-image the loacation of the image layer by some argument (may be `--image-path`).
> 
> What's your option? @imeoer

We may discuss offline. I hope we can enhance nydus-image to directly handle remote image, the same to nydusd.

**Comment by jiangliu** [comment]
Does this work for you?
```
nydus-image unpack -C nydusd.conf --outputb image.tar.gz path/to/bootstrap.file
```

File: api/src/config.rs [hunk 1]
```diff
@@ -250,6 +250,7 @@ pub struct BackendConfigV2 {
     /// Configuration for container registry backend.
     pub registry: Option<RegistryConfig>,
     /// Configuration for local http proxy.
+    #[serde(rename = "http-proxy")]
     pub http_proxy: Option<HttpProxyConfig>,
```
Comment (jiangliu) [inline api/src/config.rs:254]: This deserves a dedicated PR:)

File: src/bin/nydus-image/main.rs [hunk 2]
```diff
@@ -17,16 +17,19 @@ extern crate lazy_static;
 use std::fs::{self, metadata, DirEntry, File, OpenOptions};
 use std::os::unix::fs::FileTypeExt;
 use std::path::{Path, PathBuf};
+use std::result::Result::Ok;
```
Comment (jiangliu) [inline src/bin/nydus-image/main.rs:20]: This may be unnecessary.

Comment (loheagn) [reply src/bin/nydus-image/main.rs:20]: Yes. Fixed :)

File: api/src/config.rs (hunk 1 above)
Comment (loheagn) [reply api/src/config.rs]: Here's the pr https://github.com/dragonflyoss/image-service/pull/1086 .

**Comment by loheagn** [comment]
> Does this work for you?
> 
> ```
> nydus-image unpack -C nydusd.conf --outputb image.tar.gz path/to/bootstrap.file
> ```

NO. The argument `--bootstrap` is required.

File: src/bin/nydus-image/main.rs [hunk 3]
```diff
@@ -1210,6 +1231,20 @@ impl Command {
...
+            .context("missing backend-config argument")?;
+        let cfg = ConfigV2::from_file(cfg_file)?;
+        let backend_cfg = cfg.get_backend_config()?;
+        let backend = BlobFactory::new_backend(backend_cfg, blob_id)?;
+
+        Ok(backend)
+    }
+
     fn get_blob_id(matches: &ArgMatches) -> Result<String> {
         let mut blob_id = String::new();
```
[the remark this answers is not in the document]
Comment (loheagn) [reply src/bin/nydus-image/main.rs:1250]: You are so smart 💯

**Comment by jiangliu** [comment]
> > Does this work for you?
> > ```
> > nydus-image unpack -C nydusd.conf --outputb image.tar.gz path/to/bootstrap.file
> > ```
> 
> NO. The argument `--bootstrap` is required.

With nydus 2.2, the "--bootstrap" option for `nydus-image unpack` should be optional:)
I simplified it by making "--bootstrap" optional.

**Review by imeoer — APPROVED** [review]
